### PR TITLE
Show how to use a custom policy in output caching

### LINKED ISSUE #27687: Show how to use a custom policy in output caching
By default, output caching follows these rules:

* Only HTTP 200 responses are cached.
* Only HTTP GET or HEAD requests are cached.
* Responses that set cookies aren't cached.
* Responses to authenticated requests aren't cached.

Show how to override these rules by creating a custom policy.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: cb423d54-ee93-6dfd-87bb-ed79dff8b7b6
* Version Independent ID: cb423d54-ee93-6dfd-87bb-ed79dff8b7b6
* Content: [Output caching middleware in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/output?view=aspnetcore-7.0)
* Content Source: [aspnetcore/performance/caching/output.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/performance/caching/output.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-performance**
* GitHub Login: @tdykstra
* Microsoft Alias: **riande**

---
[Associated WorkItem - 41137](https://dev.azure.com/msft-skilling/Content/_workitems/edit/41137)

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #27687 

[Internal review URL](https://review.learn.microsoft.com/en-us/aspnet/core/performance/caching/output?view=aspnetcore-7.0&branch=pr-en-us-27688)

### CODE REVIEW

**Comment by tdykstra** [comment]
ping @Rick-Anderson @sebastienros

File: aspnetcore/performance/caching/output.md [hunk 1]
```diff
@@ -68,12 +68,33 @@ The following code applies all of the default caching rules to all of an app's e
 
 :::code language="csharp" source="output/samples/7.x/Program.cs" id="policies3":::
 
-The following code removes these defaults while applying caching to all of an app's endpoints:
+### Override the default policy
 
-:::code language="csharp" source="output/samples/7.x/Program.cs" id="policies4":::
+The following code shows how to override the default rules. The highlighted lines in the following custom policy code enable caching for HTTP Post methods and HTTP 301 responses:
```
Comment (Rick-Anderson) [inline aspnetcore/performance/caching/output.md:73]:
```suggestion
The following code shows how to override the default rules. The highlighted lines in the following custom policy code enable caching for HTTP POST methods and HTTP 301 responses:
```

File: aspnetcore/performance/caching/output.md [hunk 2]
```diff
@@ -68,12 +68,33 @@ The following code applies all of the default caching rules to all of an app's e
...
-You can override these defaults.
+To use this custom policy, create a named policy:
+
+:::code language="csharp" source="output/samples/7.x/Program.cs" id="policies3b":::
+
+And select the named policy for an endpoint:
+
+:::code language="csharp" source="output/samples/7.x/Program.cs" id="post":::
+
+### An alternative way to override the default policy
```
Comment (Rick-Anderson) [inline aspnetcore/performance/caching/output.md:85]:
```suggestion
### Alternative default policy override 
```

File: aspnetcore/performance/caching/output.md [hunk 3]
```diff
@@ -68,12 +68,33 @@ The following code applies all of the default caching rules to all of an app's e
...
+
+The custom policy class shown earlier has a private constructor and `Instance` property that creates a singleton. As an alternative, you can use Dependency Injection (DI) to initialize an instance. Give the custom policy class a public constructor instead of a private constructor, and eliminate the `Instance` property. For example:
+
+:::code language="csharp" source="output/samples/7.x/MyCustomPolicy2.cs" id="fordi":::
+
+The remainder of the class is the same as shown earlier. Add the custom policy as shown in the following example:
```
Comment (Rick-Anderson) [inline aspnetcore/performance/caching/output.md:91]:
```suggestion
The remainder of the class is the same as shown previously. Add the custom policy as shown in the following example:
```

File: aspnetcore/performance/caching/output.md [hunk 4]
```diff
@@ -68,12 +68,33 @@ The following code applies all of the default caching rules to all of an app's e
...
+
+:::code language="csharp" source="output/samples/7.x/Program.cs" id="policies3c":::
+
+With this code, DI creates the instance of the custom policy class. Any public arguments in the constructor would also be resolved.
```
Comment (Rick-Anderson) [inline aspnetcore/performance/caching/output.md:95]:
```suggestion
Using the preceding code, DI creates the instance of the custom policy class. Any public arguments in the constructor are resolved.
```

File: aspnetcore/performance/caching/output.md [hunk 5]
```diff
@@ -68,12 +68,33 @@ The following code applies all of the default caching rules to all of an app's e
...
+
+When using a custom policy as a base policy, don't call `OutputCache()` on any endpoint that the base policy should apply to. Calling `OutputCache()` adds the default policy to the endpoint.
```
Comment (Rick-Anderson) [inline aspnetcore/performance/caching/output.md:97]:
```suggestion
When using a custom policy as a base policy, don't call `OutputCache` on any endpoint that the base policy should apply to. Calling `OutputCache` adds the default policy to the endpoint.
```

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/performance/caching/output/samples/7.x/Program.cs [hunk 6]
```diff
@@ -44,6 +47,23 @@ public static void Main(string[] args)
...
+            options.AddPolicy("CachePost", MyCustomPolicy.Instance);
+        });
+        //</policies3b>
+#endif
+#if Version3c
+        //<policies3c>
+        builder.Services.AddOutputCache(options =>
+        {
+            options.AddBasePolicy(builder => 
+                builder.AddPolicy<MyCustomPolicy2>(),true);
```
Comment (sebastienros) [inline aspnetcore/performance/caching/output/samples/7.x/Program.cs:63]:
```suggestion
                builder.AddPolicy<MyCustomPolicy2>(), true);
```

**Review by sebastienros — APPROVED** [review]

**Pull request merged**